### ISSUE TITLE
Analytics middleware (Piwik?)

### ISSUE THREAD

**Issue by DenBeke** [mholt/caddy#815]
I read about a desired Piwik middleware on [Twitter](https://twitter.com/mholt6/status/729780033112301568). Maybe this is the right place to discuss what's the middleware should do... @FiloSottile @mholt 

I'm a hardcore Piwik user for years now (using the default Javascript tracking method), so enlighten me about what you think the middleware should take care of. Maybe I'll implement the middleware, because it interests me :)

**Comment by mholt** [comment]
I have to admit I'm not sure on the technical details of _how_ it would work. I've always thought it would be useful to have a self-hosted analytics service (or at least one where measurements originate on the host, rather than some remote analytics service that often gets blocked by ad blockers). How great for it to be built into a web server! (As an add-on, of course.)

It doesn't _have_ to be Piwik; it could be some other analytics service written in Go. But Piwik is a pretty comprehensive solution. It would be nice to say something in the Caddyfile like:

```
piwik ...
```

and have a working piwik install embedded into my site. Could even inject the img tag just before `</body>` or something.

**Comment by captncraig** [comment]
nothing so far afaik has actually parsed the html. It could be useful for all kinds of things, from injecting tracking scripts to rewriting links on the fly. Is there a good go package for streaming parsing of html?

**Comment by DenBeke** [comment]
Hosting Piwik itself seems way out of scope for a middleware, since it's just some simple PHP webapp.

What the middleware could do, is using Piwik's [Tracking API](http://developer.piwik.org/api-reference/tracking-api) and directly inject the new statistics into a running Piwik instance/site.
It's just an HTTP request with some data attached to it, afaik.

    piwik <piwik_instance_url> <website_id>

**Comment by DenBeke** [comment]
Embedding (i.e. alterning) Piwik tracking code into existing HTML responses seems bad behaviour to me...

**Comment by mholt** [comment]
Yeah, okay, maybe it wasn't a great idea; still, I think the idea of a self-hosted analytics tool that isn't blocked by ad blockers would be useful for site owners.

I can close this now, as I don't have a specific proposal yet.

**Comment by DenBeke** [comment]
Maybe I'll come up with something next month. Because Piwik integration in a semantically correct way would be a killer feature for people offering webhosting (and others :))...
